Ballerina-Training/Ballerina-Forum-Application/back-end/bal-forum/main.bal

```diff
@@ -0,0 +1,222 @@
+        });
+        if sentiment.label == "neg" {
+            Forbidden forbidden = {
+                body: {
+                    error_message: "Post rejected due to negative sentiment"
+                }
+            };
+            return forbidden;
+        }
+
+        UserDb userDb = check forumDbClient->queryRow(`
+            SELECT name, email, id, subscribtions FROM users WHERE id = ${id}
+        `);
+
+        User user = check transformUserFromDatabase(userDb);
+        Post post = check createPostFromNewPost(newPost, user.name);
+
+        _ = check forumDbClient->execute(`
+            INSERT INTO posts VALUES (${post.id}, ${post.title}, ${post.description}, ${post.username}, ${post.likes.toJsonString()}, ${post.comments.toJsonString()}, ${post.postedAt})
+        `);
+
+        if user.subscribtions.indexOf(post.id) is () {
+            user.subscribtions.push(post.id);
+            _ = check forumDbClient->execute(`
+                UPDATE users SET subscribtions = ${user.subscribtions.toJsonString()} WHERE id = ${id}
+            `);
+        }
+
+        Created created = {
+            body: {
+                message: "Post created successfully!"
+            }
+        };
+        return created;
+    }
+
+    resource function get posts() returns Post[]|error {
+        stream<PostDb, sql:Error?> postStream = forumDbClient->query(`SELECT * FROM posts`);
+        Post[] posts = check from var post in postStream
+            select check transformPostFromDatabase(post);
+        return posts;
+    }
+
+    resource function post posts/[string id]/like(Like like) returns Ok|NotFound|Conflict|error {
+        PostDb|error postDb = forumDbClient->queryRow(`SELECT * FROM posts WHERE id = ${id}`);
+        if postDb is PostDb {
+            Post post = check transformPostFromDatabase(postDb);
+            if post.likes.indexOf(like.userId) is int {
+                Conflict alreadyLiked = {
+                    body: {
+                        error_message: "You have already liked this post"
+                    }
+                };
+                return alreadyLiked;
+            }
+
+            post.likes.push(like.userId);
+            _ = check forumDbClient->execute(`
+                UPDATE posts SET likes = ${post.likes.toJsonString()} WHERE id = ${id}
+            `);
+
+            Ok liked = {
+                body: {
+                    message: "Post liked successfully!"
+                }
+            };
+            return liked;
+        }
+
+        NotFound postNotFound = {
+            body: {
+                error_message: "Post not found"
+            }
+        };
+        return postNotFound;
+    }
+
+    resource function get posts/[string id]() returns Post|NotFound|error {
+        PostDb|error postDb = forumDbClient->queryRow(`SELECT * FROM posts WHERE id = ${id}`);
+        if postDb is PostDb {
+            Post post = check transformPostFromDatabase(postDb);
```

Comment:
Do we need the variable `post`?

Reply:
No, will change it according to the next comment

Reply:
Fixed via [1eae7ab](https://github.com/ballerina-guides/training-content/pull/12/commits/1eae7abbc8d1d6bdd3a677b4fac348875ee53fb7)